cs/src/Connections/TunnelClient.cs

```diff
@@ -221,19 +225,62 @@ protected async Task StartSshSessionAsync(Stream stream, CancellationToken cance
 
     private void ConfigurePortForwardingService()
     {
-        if (SshPortForwardingService is PortForwardingService pfs)
+        var pfs = SshPortForwardingService;
+        if (pfs == null)
         {
-            pfs.AcceptLocalConnectionsForForwardedPorts = this.acceptLocalConnectionsForForwardedPorts;
-            if (pfs.AcceptLocalConnectionsForForwardedPorts)
-            {
-                pfs.TcpListenerFactory = new RetryTcpListenerFactory(this.localForwardingHostAddress);
-            }
+            return;
+        }
+
+        pfs.AcceptLocalConnectionsForForwardedPorts = this.acceptLocalConnectionsForForwardedPorts;
+        if (pfs.AcceptLocalConnectionsForForwardedPorts)
+        {
+            pfs.TcpListenerFactory = new RetryTcpListenerFactory(this.localForwardingHostAddress);
+        }
+
+        if (ConnectionProtocol == TunnelRelayTunnelClient.WebSocketSubProtocolV2)
+        {
+            pfs.MessageFactory = this;
+            pfs.ForwardedPortConnecting += OnForwardedPortConnecting;
+        }
+    }
+
+    /// <summary>
+    /// Invoked when a forwarded port is connecting.
+    /// </summary>
+    protected virtual void OnForwardedPortConnecting(
+        object? sender, ForwardedPortConnectingEventArgs e)
+    {
+        var channel = e.Stream.Channel;
+        var relayResponseMessage = channel.OpenConfirmationMessage
+            .ConvertTo<PortRelayConnectResponseMessage>();
```

Comment:
How is this going to work during the transition period, when the open confirmation message is not yet actually `PortRelayConnectResponseMessage` (i.e. the service still sends out the old message) ? Will it cause binary protocol incompatibility?

Reply:
The `OnForwardedPortConnecting()` method is only hooked up for V2 protocol. So it's a breaking change for V2, but nobody is using that version yet.

I'll add more comments clarifying that.